src/backend/services/auth/jwt.py

```diff
@@ -64,8 +99,77 @@ def decode_jwt(self, token: str) -> dict:
             )
             return decoded_payload
         except jwt.ExpiredSignatureError:
-            logger.warning("Token has expired.")
-            return None
+            logger.warning("JWT Token has expired.")
+            decoded_payload = jwt.decode(
+                token,
+                self.secret_key,
+                algorithms=[self.ALGORITHM],
+                options={"verify_exp": False},
+            )
+            return decoded_payload
         except jwt.InvalidTokenError:
-            logger.warning("Invalid token.")
+            logger.warning("JWT Token is invalid.")
             return None
+
+    @staticmethod
+    def check_validity(payload: dict, session: Session = Depends(get_session)) -> str:
+        """
+        Checks a given JWT payload for its validity.
+
+        Args:
+            payload (dict): JWT payload.
+
+        Returns:
+            str: One of the following strings depending on the validity:
+                - "valid": Token is valid.
+                - "refreshable": Token is valid and within the refresh availability window.
+                - "expired": Token is expired or blacklisted.
+                - "invalid": Token is invalid.
+        """
+
+        if payload is None or any(
+            [
+                "context" not in payload,
+                "jti" not in payload,
+                "exp" not in payload,
+                "strategy" not in payload,
+                "iat" not in payload,
+            ]
+        ):
+            return "invalid"
+
+        now = datetime.datetime.now(datetime.timezone.utc)
+
+        # Check if token is blacklisted
+        blacklist = (
+            session.query(Blacklist)
+            .filter(Blacklist.token_id == payload["jti"])
+            .first()
+        )
+
+        if blacklist is not None:
+            logger.warning("JWT payload is blacklisted.")
+            return "expired"
+
+        # Check if token is expired; either we're past the expiry time or iat is more than 3 months ago
```

Comment:
I'm using `iat` to mark the time that the user logs in. When we generate a new JWT using the refresh logic, we keep the `iat` from the original JWT. This lets us eventually force the user to log in again instead of refreshing their auth token forever.